### PR TITLE
Webpack5

### PR DESCRIPTION
**Pull request by barisusakli**
### Core plugin pull requests
- nodebb-plugin-2factor (PR https://github.com/NodeBB/nodebb-plugin-2factor/pull/73)
- nodebb-plugin-composer-default (PR https://github.com/NodeBB/nodebb-plugin-composer-default/pull/157)
- nodebb-plugin-dbsearch (**works without changes**)
- nodebb-plugin-emoji (PR https://github.com/NodeBB/nodebb-plugin-emoji/pull/53)
- nodebb-plugin-emoji-android (**works without changes**)
- nodebb-plugin-markdown (PR https://github.com/julianlam/nodebb-plugin-markdown/pull/209)
- nodebb-plugin-mentions (**works without changes**)
- nodebb-plugin-spam-be-gone (PR https://github.com/akhoury/nodebb-plugin-spam-be-gone/pull/87)
- nodebb-rewards-essentials (**works without changes**)
- nodebb-theme-lavender (PR https://github.com/NodeBB/nodebb-theme-lavender/pull/93)
- nodebb-theme-persona (PR https://github.com/NodeBB/nodebb-theme-persona/pull/545)
- nodebb-theme-slick (PR https://github.com/pichalite/nodebb-theme-slick/pull/59)
- nodebb-theme-vanilla (**works without changes**)
- nodebb-widget-essentials (PR https://github.com/NodeBB/nodebb-widget-essentials/pull/60)
  

### Other plugins
 - nodebb-plugin-quickstart (**works without changes**)
 - nodebb-plugin-browsing-users (**works without changes**)
 - nodebb-plugin-feed (**works without changes**)
 - nodebb-plugin-glossary (**works without changes**)
 - nodebb-plugin-recent-cards (**works without changes**)
 - nodebb-plugin-custom-pages (**works without changes**)
 - nodebb-plugin-tenor-gif (**works without changes**)
 - nodebb-plugin-blog-comments (**works without changes**)
 - nodebb-plugin-canned-responses (**works without changes**)
 - nodebb-plugin-question-and-answer (**works without changes**)
 - nodebb-plugin-google-analytics (**works without changes**)
 - nodebb-plugin-emailer-sendgrid (**works without changes**)
 - nodebb-plugin-sso-google (**works without changes**)
 - nodebb-plugin-sso-facebook (**works without changes**)
 - nodebb-plugin-sso-github (**works without changes**)
 - nodebb-plugin-sso-twitter (**works without changes**)
 - nodebb-plugin-iframely (**works without changes**)
 - nodebb-plugin-imgur (**works without changes**)
 - nodebb-plugin-registration-question  (**works without changes**)
 - nodebb-plugin-github-embed (**works without changes**)
 - nodebb-plugin-osm-map (PR https://github.com/NodeBB-Community/nodebb-plugin-osm-map/pull/17)
 - nodebb-plugin-session-sharing (**works without changes**)
 - nodebb-plugin-discord-notification (PR https://github.com/amargon/nodebb-plugin-discord-notification/pull/12)

  
## Breaking Changes &rarr; [See the migration guide](https://nodebb.org/blog/migration-guide-for-v2/)

### CODE REVIEW

**Comment by pitaj** [comment]
Why do we need app.importScript? Doesn't webpack support handling requirejs imports? Why not add some aliases in webpack for chart, jquery-ui, etc? Why not use webpack `import()` for the templates? Webpack should be able to handle the module.exports wrapping.

**Comment by barisusakli** [comment]
Putting `require(['moduleName'], function (module){})` doesn't work when it is put in a widget or custom js as they are executed at run time and not during webpack build.

Yeah I might add the aliases for chart.js and the others to reduce # of files changed in this pr. Using webpack import() for templates results in hundreds of files being built during `./nodebb build` so just requesting them dynamically.

**Comment by pitaj** [comment]
> doesn't work when it is put in a widget or custom js as they are executed at run time and not during webpack

Ah, widgets and custom js okay.

> Using webpack import() for templates results in hundreds of files being built during ./nodebb build so just requesting them dynamically.

I'd much rather have hundreds of extra files than custom script importing just for templates.

File: webpack.common.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+module.exports = {
+	plugins: [
+		new CleanWebpackPlugin(), // cleans dist folder
+	],
+	entry: {
+		nodebb: './build/public/src/client.js',
+		admin: './build/public/src/admin/admin.js',
+	},
+	output: {
+		filename: '[name].min.js',
```
Comment (pitaj) [inline webpack.common.js:30]: Would recommend adding a hash to the filename for cache purposes.

Comment (barisusakli) [reply webpack.common.js:30]: https://github.com/NodeBB/NodeBB/commit/e190ff875acf97a6eaa513171ff78819fb4b4cb0

**Pull request merged**